# Assignment 4.1

> Replace all TODOs with your code.
>
> Do not change any other code and do not add/remove cells!

## Sets

### Task 1

Create an empty set called 'my_set'.

In [ ]:
my_set = set()

print(my_set)  # should print set()

set()


### Task 2

Let's turn a given sentence into a set of words.


In [ ]:
sentence = "abc cba acd abc"

words = set(sentence.split(' '))

print(words)  # should be {'acd', 'cba', 'abc'} or in other order

{'acd', 'cba', 'abc'}


### Task 3

Now, let's remove from `words` all the words that are present in `sentence2`.

In [4]:
sentence2 = 'cba baa acc'

words = set(sentence2.split())
words.remove('acc')

print(words) # should print {'acd', 'abc'} or {'abc', 'acd'}

{'baa', 'cba'}


### Task 4

Let's now play around with statistics and probability!

Generate `N` random numbers in the range `[1, N]` and keep track of a percent of unique numbers in the result.
Play around with different values of `N`; how does it affect the percentage of unique numbers? How do you think why?

In [ ]:
from random import randint

N = 10000000

def generate_unique_numbers(N):
  unique_numbers = set()
  for number in range(N):
        random_number = randint(1, N)
        unique_numbers.add(random_number)

  percent_unique = (len(unique_numbers) / N) * 100
  return percent_unique


percent_unique_numbers = generate_unique_numbers(N)
print(f"Процент уникальных чисел при N = {N}: {percent_unique_numbers:.2f}%")

Процент уникальных чисел при N = 10000000: 63.21%


For math geeks: if you set rather big `N`, you see that the percentage is slowly approaching ~63.21. It is a modification of a [Birthday problem](https://en.wikipedia.org/wiki/Birthday_problem), but a simpler and more relevant explanation is [here](https://math.stackexchange.com/questions/5775/how-many-bins-do-random-numbers-fill).


## Dictionaries

### Task 5

You are given a matrix `m`. Return a distinct list containing all the values that are present in at least `n` rows.

Hint: there can be duplicates in rows; perhaps we would want to eliminate them - how?

In [5]:
m = [
    [1, 1, 2, 3],
    [1, 2, 6, 7],
    [1, 7, 9, 9]
]

n = 2
res = []

for i in range(len(m) - 1):

  for part in m[i]:
    if part  in m[i + 1]:
      res.append(part)

res = list(set(res))
print(res)  # should print [1, 2, 7]

[1, 2, 7]
